### DISCUSSION TITLE
Feature Request: Automatic Series Sequence Parsing in Library Scan

### DISCUSSION THREAD

**Discussion by AndryXY** [advplyr/audiobookshelf#3519] [Ideas]
Hello everyone,

Currently, the series tag can be added automatically when scanning the library using the following folder structure:
[author] -> [series] -> [Vol.<sequence> - title].

It would be great if the sequence number could also be parsed from the folder structure and added to the metadata, not just the series name. The sequence number is already identified from the folder naming convention, so it would make sense to automatically include it in the metadata.

![sequence_2](https://github.com/user-attachments/assets/4b7a03e9-7943-472c-ac7d-af2e16b9fa2f)

Is it possible to implement this feature to support both the series name and sequence number during the library scan?

**Comment by CLHatch** [comment]
> Hello everyone,
> 
> Currently, the series tag can be added automatically when scanning the library using the following folder structure: [author] -> [series] -> [Vol. - title].
> 
> It would be great if the sequence number could also be parsed from the folder structure and added to the metadata, not just the series name. The sequence number is already identified from the folder naming convention, so it would make sense to automatically include it in the metadata.
> 
> ![sequence_2](https://private-user-images.githubusercontent.com/1040136/376385386-4b7a03e9-7943-472c-ac7d-af2e16b9fa2f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg5NDE3OTMsIm5iZiI6MTcyODk0MTQ5MywicGF0aCI6Ii8xMDQwMTM2LzM3NjM4NTM4Ni00YjdhMDNlOS03OTQzLTQ3MmMtYWM3ZC1hZjJlMTZiOWZhMmYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAxNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMTRUMjEzMTMzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YzI3MmU3MGIyNTIxNmJkYzMxYTFiNWZhMjU4YzhhYzZmMTMxMTRjMGM5ZjNhODExNDY4NGZkYjBkMzUwYmM2YiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.BePtz8fYtCPHSPUmQA_5J0sw2yYR3IId9BjuMeWQLVo)
> 
> Is it possible to implement this feature to support both the series name and sequence number during the library scan?

This is already a feature. There are a few formats that work, but the one I use is including the sequence number at the beginning of the folder name. 

This is the naming format I use:

`Author One; Author Two/Series Name/## - (YYYY) - Long Title [ASIN]/Short Title [ASIN].ext`

The ASIN is the Audible ID if it's an Audible book.

 He's an example from my library:
![image](https://github.com/user-attachments/assets/f6b2d0ff-9553-4b09-b3ca-16691b478baf)

Edit:
Here's where it explains the options for the naming. https://www.audiobookshelf.org/docs/#book-title-folder-naming

**Reply by AndryXY** [reply]
Thank you very much for your feedback. I've already tried several variants, but I'm not sure whether I've tested yours correctly. You always have to make sure that the metadata.json is deleted, otherwise the wrong data will be taken from there when re-importing. I'll test that in a few days. In the above example, I use the following folder path (maybe helpfull to others):

- audiobooks/Enid Blyton/Fünf Freunde/Fünf Freunde - Vol.02 - im Wanderzirkus/[files]

Unfortunately this does not work until version 2.14.0

**Reply by CLHatch** [reply]
> Thank you very much for your feedback. I've already tried several variants, but I'm not sure whether I've tested yours correctly. You always have to make sure that the metadata.json is deleted, otherwise the wrong data will be taken from there when re-importing. I'll test that in a few days. In the above example, I use the following folder path (maybe helpfull to others):
> 
> * audiobooks/Enid Blyton/Five Friends/Five Friends - Vol.02 - at camp/[files]
> 
> Unfortunately this does not work until version 2.14.0

I think what's broken with your pattern is no space after the `Vol.`.  Try with a space: `audiobooks/Enid Blyton/Five Friends/Five Friends - Vol. 02 - at camp/[files]` 

Edit:
However, might give my scheme a try as well.  With the sequence number at the beginning of the folder (with leading zeroes for longer series), the books are sorted correctly in the folder.

Edit 2:
Ah, my mistake.  Yours would sort correctly also.  It's just that my scheme makes it more obvious under the "Authors" folder if something is a series or a book.
![image](https://github.com/user-attachments/assets/81032754-fa1d-4037-a5b2-1c3adc78df10)

**Reply by AndryXY** [reply]
You are a genius! Thank you - I would have had a look at the code. Unfortunately, this important information is missing in the documentation: 

[Quote: https://www.audiobookshelf.org/docs/#book-title-folder-naming]
- The sequence can be anywhere in the folder name.
- It must start with “ - ‘ or ’. ”
- If it is not at the beginning of the folder name, it must be followed by “ - ‘ and ’Vol” “Vol.” “Volume” or ”Book”

It would be great if someone could adjust the documentation here. I will leave the discussion open until then.

**Reply by CLHatch** [reply]
> You are a genius! Thank you - I would have had a look at the code. Unfortunately, this important information is missing in the documentation:
> 
> [Quote: https://www.audiobookshelf.org/docs/#book-title-folder-naming]
> 
> * The sequence can be anywhere in the folder name.
> * It must start with “ - ‘ or ’. ”
> * If it is not at the beginning of the folder name, it must be followed by “ - ‘ and ’Vol” “Vol.” “Volume” or ”Book”
> 
> It would be great if someone could adjust the documentation here. I will leave the discussion open until then.

Good to know that fixed it.  Was more an educated guess, given the examples all had spaces before the number, even though it didn't specifically say the space was required.

**Reply by advplyr** [reply]
This is specified in the docs. The spaces are in the quotes " - " and ". "

**Reply by CLHatch** [reply]
> This is specified in the docs. The spaces are in the quotes " - " and ". "

I think where the confusion is is that it's not in the quotes in `“ - ‘ and ’Vol” “Vol.” “Volume” or ”Book”`, other than the first one.  Just noticed looking at that that there are a couple of single quotes mixed in with the double  quotes, btw.
